### DISCUSSION TITLE
Transform data returned by a Query

### DISCUSSION THREAD

**Discussion by ndlang** [Budibase/budibase#3904] [Help] [answered]
I am attempting to transform data returned by a query (MySQL). I am basically trying to group by the member name. this should be come more clear after the rest of the post.
Query:
```
SELECT 
	CONCAT(im.first_name,' ', im.last_name) as 'member_name',
	cr.paid,
	cr.banquet,
	cr.sitin,
	cr.challenge_exam,
	oc.title 
FROM course_registration cr
	INNER JOIN imsa_member im ON im.member_id = cr.imsa_member_id
	INNER JOIN offered_course oc on oc.offered_course_id =cr.offered_course_id
;
```
Transformer js code:
```
class Member_Registrations {
	constructor(name,paid,banquet,sitin,courses) {
		this.name = name;
		this.paid = paid;
		this.banquet = banquet;
		this.courses = [];
	}
}

const raw_regs = data

let f_regs = []
for (let reg of raw_regs){
	let result = f_regs.findIndex(item => item.name == reg.member_name)
	if (result = -1) {
		let m = new Member_Registrations(
			reg.member_name, 
			reg.paid,
			reg.banquet,
			reg.sitin
		)
		m.courses.push(reg.title)
		
		f_regs.push(m)
		
	} else {
		f_regs[result].courses.push(reg.title)
	}
}

return JSON.stringify(f_regs)
```
Result:
```
{
  "value": "[{\"name\":\"Jim Fitzgerald\",\"paid\":null,\"banquet\":null,\"courses\":[\"Work Zone Traffic Control\"]},{\"name\":\"Ted Raney\",\"paid\":null,\"banquet\":null,\"courses\":[\"Work Zone Traffic Control\"]},{\"name\":\"Bill Bowen\",\"paid\":1,\"banquet\":null,\"courses\":[\"Work Zone Traffic Control\"]},{\"name\":\"Richard Dantoni\",\"paid\":null,\"banquet\":null,\"courses\":[\"Work Zone Traffic Control\"]},{\"name\":\"Richard Dantoni\",\"paid\":null,\"banquet\":null,\"courses\":[\"Work Zone Traffic Control\"]},...
}
```

I am trying to get the formate for each entry as shown below:
```
{
    name: name,
    paid: true
    banquet: true
    sitin: true
   challenge_exam:true
   courses: [course 1, course 2]
}
```
Any Ideas?

**Comment by mjashanks** [comment, marked as answer]
you're almost there - your issue is with the "equals" comparison....

`let result = f_regs.findIndex(item => item.name == reg.member_name)`
should be
`let result = f_regs.findIndex(item => item.name === reg.member_name)`

and

`if (result = -1) {`
should be
`if (result === -1) {`

i.e. use `===` when doing an equals comparison in javascript.

**Reply by ndlang** [reply]
That did the trick. Thank you for your help.

**Reply by ndlang** [reply]
Another Quick Question, I am having some trouble getting the return format correct. I was reading the docs and tried to follow the example but I apparently am not understanding something:

JS code with adjustments from above:
//return data

class Member_Registrations {
	constructor(name,paid,banquet,sitin,courses) {
		this.name = name;
		this.paid = paid;
		this.banquet = banquet;
		this.courses = [];
	}
}


const raw_regs = data

let f_regs = []
for (let reg of raw_regs){
	let result = f_regs.findIndex(item => item.name === reg.member_name)
	if (result === -1) {
		let m = new Member_Registrations(
			reg.member_name, 
			reg.paid,
			reg.banquet,
			reg.sitin
		)
		m.courses.push(reg.title)
		
		f_regs.push({...m})
		
	} else {
		f_regs[result].courses.push(reg.title)
	}
}

//return JSON.stringify(f_regs)

return Object.entries(f_regs)

//const entries = Object.entries(f_regs)
//return entries.map(([name, paid, banquet, courses]) => ({ name, paid, banquet, courses }))```

Output:
[
  "0",
  {
    "name": "Jim Fitzgerald",
    "paid": null,
    "banquet": null,
    "courses": [
      "Work Zone Traffic Control",
      "Work Zone Traffic Control",
      "Level I Traffic Signal Field"
    ]
  }
]

Adjustment attempt 1:
...
return Object.entries(f_regs)

//const entries = Object.entries(f_regs)
//return entries.map(([name, paid, banquet, courses]) => ({ name, paid, banquet, courses }))```

Output:
{
  "name": "0",
  "paid": {
    "name": "Jim Fitzgerald",
    "paid": null,
    "banquet": null,
    "courses": [
      "Work Zone Traffic Control",
      "Work Zone Traffic Control",
      "Level I Traffic Signal Field"
    ]
  }
}

**Reply by ndlang** [reply]
Disregard the last message I sent I figured it out.

`...
return f_regs`
